### DISCUSSION TITLE
Better support for podman?

### DISCUSSION THREAD

**Discussion by LewisGaul** [gabrieldemarmiesse/python-on-whales#516] [General]
As far as I understand it, POW does support podman, in a kind of best-effort way. [1]

I may be able to take on some work to improve podman support if there's appetite for this. Following my mention in https://github.com/gabrieldemarmiesse/python-on-whales/discussions/515, some things that are currently on my mind:
1. Lack of `podman pod ...` support (https://docs.podman.io/en/stable/markdown/podman-pod.1.html)
   * May also be supported with docker using [this plugin](https://github.com/nicholasdille/docker-pod) (I haven't tried it)
1. Other CLI compatibility I'm likely to flush out in the coming months
1. Run tests using podman in CI (https://github.com/gabrieldemarmiesse/python-on-whales/issues/353)
1. Documentation could be made less focused on docker and more on "containers" in general
1. API names could use "container" (or "ctr") instead of "docker"
   * I find myself doing this manually, e.g. [here](https://github.com/ios-xr/xrd-ha-app/blob/d740441293304d6fc42a71b2a970132eec26a5c3/tests/it/conftest.py#L21)
   * Appreciate this isn't ideal with backwards compatibility... would have to provide aliases rather than removing the old names

If there's interest in some/all of these points I'd be happy to create issues for tracking, and hopefully get time to make some of the changes at some point.

---
[1]

From https://github.com/gabrieldemarmiesse/python-on-whales/issues/408#issuecomment-1413778388:
> I am open to support options that are present in certain implementations of Docker and not others. Though I will limit it to
> * Docker
> * Podman
> * Nerdctl

---

Past podman issues/PRs:
1. https://github.com/gabrieldemarmiesse/python-on-whales/issues/349
1. https://github.com/gabrieldemarmiesse/python-on-whales/pull/350
1. https://github.com/gabrieldemarmiesse/python-on-whales/issues/373
1. https://github.com/gabrieldemarmiesse/python-on-whales/issues/408
1. https://github.com/gabrieldemarmiesse/python-on-whales/issues/472

**Comment by gabrieldemarmiesse** [comment]
I would be happy to support more podman-only commands, BUT we will need first to run the unit tests with podman in the CI. If it's not the first thing we do, we might end up with broken features.

Once the unit tests are running, we can go ham with the podman functions, sure thing, it's not hard to maintain (python-on-whales is, in general, not very time-consuming to maintain since we don't add much logic).

Concerning the documentation and the API: Docker will and should be a first-class citizen. It will stay like that unless numbers show that the majority of container users have migrated to another tool. However, as long as it doesn't impact Docker users, we are free to add pages to the documentation, guides, new apis, etc...

**Reply by LewisGaul** [reply]
> we will need first to run the unit tests with podman in the CI. If it's not the first thing we do, we might end up with broken features.

Agreed this makes sense. I might have a go at getting this working sometime soon.

> Docker will and should be a first-class citizen. It will stay like that unless numbers show that the majority of container users have migrated to another tool.

If it were me the balance would be more like "when the number of users are of the same order of magnitude" rather than when the majority have moved away from docker, however this is obviously quite hard to judge. I think there are domains where podman will have a *much* higher use than docker (e.g. where podman's support for running as non-root user is required, users on RHEL, perhaps users of Alpine although that's a bit niche), and there are domains in which people won't have even heard of podman (people casually trying out containers, e.g. on a personal machine). The thing is that the "container" naming is suitable for all users, whereas the "docker" naming is misleading if podman is actually desired/in use.

That being said, the idea of changing APIs or adding aliases isn't particularly appealing, I'll admit. It's probably just something to solve using documentation as you suggest.

**Reply by gabrieldemarmiesse** [reply]
We can work things out as we progress. It will be easier to discuss this with specific examples to illustrate :)

**Comment by LewisGaul** [comment]
I've given some thought to adding podman to the CI, with a tidyup PR [here](https://github.com/gabrieldemarmiesse/python-on-whales/pull/517) and some thoughts on approach [here](https://github.com/gabrieldemarmiesse/python-on-whales/issues/353#issuecomment-1873542214).

I've also started working on adding support for `podman pod` CLI commands on my [branch](https://github.com/gabrieldemarmiesse/python-on-whales/compare/master...LewisGaul:podman-pod-wip).

**Comment by LewisGaul** [comment]
A related point/issue on the docker-first stance of PoW - when writing a CLI tool that wraps podman (to run on RHEL hosts) if something goes wrong the error message shown to the user is something like:
`python_on_whales.exceptions.DockerException: The docker command executed was '/bin/podman image inspect docker.io/library/ubuntu'.`

Unfortunately "The docker command executed" is wrong and misleading here, but it's difficult for us to change what's presented to the user here (at least without a horrible hack). If this were changed to "ContainerException: The container command executed ..." it would be much better.

**Reply by gabrieldemarmiesse** [reply]
We can just say "the command executed", I don't mind changing this error message if it can help podman users

**Reply by LewisGaul** [reply]
Created https://github.com/gabrieldemarmiesse/python-on-whales/pull/585 for this.
